### ISSUE TITLE
💅 noUnnecessaryConditions: false positive on useRef(...).current null-guards (regression in 2.5.6)

### ISSUE THREAD

**Issue by liorp** [biomejs/biome#11360]
### Environment information

```block
CLI:
  Version:                      2.5.8

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_DISTRIBUTION:           npm
  BIOME_CONFIG_PATH:            unset
  JS_RUNTIME_VERSION:           v26.7.0
  JS_RUNTIME_NAME:              node
```

`@types/react` 19.2.18, `react` 19.2.8.

### Rule name

`suspicious/noUnnecessaryConditions`

### Summary

Since 2.5.6, a null-check on `useRef(...).current` is reported as
`This condition is always truthy` / `The value's type can never be falsy`,
even though `current` is typed `T | null`.

The same object shape written as a plain annotation is **not** reported, so the
defect is in how the ref type coming out of React's `useRef` is resolved, not in
the rule's handling of `{ current: T | null }`.

### Reproduction

`biome.json`:

```json
{
  "linter": {
    "enabled": true,
    "domains": { "project": "recommended", "types": "recommended" },
    "rules": {
      "preset": "recommended",
      "suspicious": { "noUnnecessaryConditions": "error" }
    }
  }
}
```

`src/narrow.ts`:

```ts
import { useRef } from "react";

// A - REPORTED (line 7)
export function a(): number {
	const r = useRef<HTMLCanvasElement | null>(null);
	const v = r.current;
	if (!v) return 0;
	return v.width;
}

// B - NOT reported. Same shape, plain annotation. This one is correct.
export function b(): number {
	const r: { current: HTMLCanvasElement | null } = { current: null };
	const v = r.current;
	if (!v) return 0;
	return v.width;
}

// C - REPORTED (line 22). Guard directly on the member expression.
export function c(): number {
	const r = useRef<HTMLCanvasElement | null>(null);
	if (!r.current) return 0;
	return r.current.width;
}

// D - REPORTED (line 30). Non-null type argument, React's ref-callback style.
export function d(): number {
	const r = useRef<HTMLCanvasElement>(null);
	const v = r.current;
	if (!v) return 0;
	return v.width;
}
```

Output on 2.5.8:

```block
src/narrow.ts:7:6  lint/suspicious/noUnnecessaryConditions  This condition is always truthy.
src/narrow.ts:22:6 lint/suspicious/noUnnecessaryConditions  This condition is always truthy.
src/narrow.ts:30:6 lint/suspicious/noUnnecessaryConditions  This condition is always truthy.
```

`b` is absent from that list. That is the useful signal: the rule reasons about
`{ current: T | null }` correctly when the type is written directly, and only
goes wrong when the value comes from `useRef`.

In `@types/react` 19.x, `useRef<HTMLCanvasElement | null>(null)` yields
`RefObject<HTMLCanvasElement | null>` whose `current` is
`HTMLCanvasElement | null`. `useRef<HTMLCanvasElement>(null)` (case D) selects the
`initialValue: T | null` overload and is also nullable. All three guards are
required.

### Version behaviour

Same file, same config, same `@types/react`. Count of
`noUnnecessaryConditions` diagnostics:

| Version | Count |
| --- | --- |
| 2.5.4 | 0 |
| 2.5.5 | 0 |
| 2.5.6 | **3** |
| 2.5.7 | **3** |
| 2.5.8 | **3** |

The regression is introduced in 2.5.6.

### Expected behaviour

None of `a`, `c`, `d` should be reported.

### Impact

`const el = someRef.current; if (!el) return;` is the standard way to read a DOM
ref inside an effect, where `current` is genuinely `null` until React attaches
the node. Acting on the diagnostic and deleting the guard produces a null
dereference at runtime. In one repository this fired on 7 files, every one of
them an ordinary DOM-ref guard in an effect or an event handler.

### Playground link

Not provided: the rule is type-aware and depends on `@types/react` being
resolvable, so a playground link risks looking clean for the wrong reason.

**Comment by liorp** [comment]
Narrowed further: **React is not involved at all.** A single generic function and
a one-property generic interface are enough, so this is generic instantiation
rather than anything about `useRef`'s overloads.

`src/noreact.ts`:

```ts
interface RefObject<T> { current: T }
declare function myRef<T>(initialValue: T): RefObject<T>;
declare function myRefOverloaded<T>(initialValue: T): RefObject<T>;
declare function myRefOverloaded<T>(initialValue: T | null): RefObject<T | null>;

type El = { width: number };

// 1 - REPORTED "always truthy". Single signature, explicit union type argument.
export function one(): number {
	const r = myRef<El | null>(null);
	const v = r.current;
	if (!v) return 0;
	return v.width;
}

// 2 - REPORTED "always truthy". Overloaded, explicit union type argument.
export function two(): number {
	const r = myRefOverloaded<El | null>(null);
	const v = r.current;
	if (!v) return 0;
	return v.width;
}

// 3 - REPORTED "always truthy". Overloaded, non-null type argument.
export function three(): number {
	const r = myRefOverloaded<El>(null);
	const v = r.current;
	if (!v) return 0;
	return v.width;
}

// 4 - REPORTED "always falsy". No explicit type argument; the literal is not widened.
export function four(): boolean {
	const r = myRef(false);
	if (r.current) return true;
	return false;
}
```

All four are reported on 2.5.8. Case 1 is the smallest: no overloads, no React,
one type parameter used in one property.

There appear to be two distinct defects here:

1. **Cases 1-3** - instantiating `RefObject<T>` with a union type argument loses
   the nullish member, so `current` is treated as non-nullable.
2. **Case 4** - inferring `T` from a literal argument keeps the literal type
   `false` instead of widening to `boolean`, so `current` is treated as always
   falsy. This is the shape behind a real-world hit of ours, a boolean re-entrancy
   guard `if (syncingRef.current) return;`.

### Verified against `tsc`

```ts
const r = myRef<El | null>(null);
export const shouldError: El = r.current;      // must error if `current` is nullable

const b = myRef(false);
export const shouldBeFine: boolean = b.current; // must compile if `current` is widened
```

`tsc --noEmit --strict` (5.9.3) gives exactly one error:

```block
src/tscheck.ts(7,14): error TS2322: Type 'El | null' is not assignable to type 'El'.
  Type 'null' is not assignable to type 'El'.
```

So TypeScript agrees `current` is `El | null` in the first case and `boolean` in
the second, which is the opposite of what the rule concludes in both.

**Comment by liorp** [comment]
Narrowed to a precise trigger. The defect needs **all three** of: a generic, a
union supplied as an explicit type argument, and the type parameter used as a
**property** type.

```ts
interface RefObject<T> { current: T }
type El = { width: number };
type MaybeEl = El | null;

declare function myRef<T>(initialValue: T): RefObject<T>;
declare function plainRef(): { current: El | null };
declare function directUnion<T>(v: T): T | null;

// v1 - REPORTED (wrong). Named alias union as the type argument.
export function v1(): number {
	const r = myRef<MaybeEl>(null);
	if (!r.current) return 0;
	return r.current.width;
}

// v2 - not reported (correct). No generic; property is already a union.
export function v2(): number {
	const r = plainRef();
	if (!r.current) return 0;
	return r.current.width;
}

// v3 - not reported (correct). Union formed in the generic's return type.
export function v3(): number {
	const x = directUnion<El>({ width: 1 });
	if (!x) return 0;
	return x.width;
}

// v4 - REPORTED (wrong). Same as v1 with `undefined`.
export function v4(): number {
	const r = myRef<El | undefined>(undefined);
	if (!r.current) return 0;
	return r.current.width;
}
```

Summary of the boundary:

| case | shape | 2.5.8 |
| --- | --- | --- |
| v1 | `myRef<MaybeEl>(null)`, `T` used as property type | reported (wrong) |
| v2 | non-generic, property typed `El \| null` | correct |
| v3 | generic, union formed in return type `T \| null` | correct |
| v4 | `myRef<El \| undefined>(undefined)` | reported (wrong) |

v2 and v3 being clean is the useful part: union handling and generic return
types are both fine on their own. It is substituting a union type argument into a
property position that drops the nullish member. Inline unions and named aliases
behave identically, and `null` and `undefined` both reproduce.

**Comment by liorp** [comment]
Root cause, from a debug build of `main` (`6cd32636`, which still reproduces):

**Explicit type arguments on a call are never lowered into the type model, so a
generic call's type parameters are inferred from the call arguments alone.**

`crates/biome_js_type_info/src/type_data.rs:1340`:

```rust
pub struct TypeofCallExpression {
    pub callee: TypeReference,
    pub arguments: Box<[CallArgumentType]>,
}
```

There is no field for type arguments, and `local_inference.rs:485` never reads
`expr.type_arguments()`:

```rust
AnyJsExpression::JsCallExpression(expr) => match expr.callee() {
    Ok(callee) => Self::from(TypeofExpression::Call(TypeofCallExpression {
        callee: collector.reference_to_resolved_expression(scope_id, &callee),
        arguments: CallArgumentType::types_from_js_call_arguments(
            collector,
            scope_id,
            expr.arguments().ok(),
        ),
    })),
```

`flattened_function_call` (`flattening/expressions.rs:1047`) then calls
`infer_generic_arguments`, which zips `function.parameters` with
`expr.arguments` only. So for `id<El | null>(null)`, `T` is bound to the type of
the argument `null` rather than the written `El | null`.

This predicts the exact verdicts observed, including the ones that look
contradictory:

| call | `T` bound to | resulting value | reported condition |
| --- | --- | --- | --- |
| `id<El \| null>(null)` | `null` | `null` | `!x` always truthy |
| `id<null \| El>(null)` | `null` | `null` | `!x` always truthy |
| `id<string \| El>("a")` | `"a"` | `"a"` | `!x` always falsy |
| `useRef<HTMLElement \| null>(null)` | `null` | `current: null` | `!current` always truthy |

The order of union members is irrelevant, which matches: the written type
argument is simply not consulted. It also explains the two clean cases from my
earlier comment - an annotated variable (`const r: RefObject<El | null> = ...`)
and an inferred type argument (`mk(v)` where `v: El | null`) both bypass this
path and behave correctly.

Minimal reproduction is now two lines:

```ts
declare function id<T>(v: T): T;
const x = id<string | null>(null);
if (!x) {} // "This condition is always truthy"
```

A fix presumably needs `TypeofCallExpression` to carry
`type_arguments: Box<[TypeReference]>`, populated with the existing
`TypeReference::types_from_ts_type_arguments` helper (`local_inference.rs:2517`,
already used for `TsReferenceType`), and `flattened_function_call` to bind those
to the function's type parameters before falling back to argument inference.

That touches an interned core type, so I did not want to push a patch shaped by
guesswork about your interning and snapshot conventions. Happy to open a PR
along those lines if that is the direction you would take, or to test a patch.

Note this is not specific to `noUnnecessaryConditions` - any type-aware rule
reading a generic call's return type is affected. I filed #11359 separately for a
`useAwaitThenable` false positive; that one has a different trigger (a
`Promise`-returning property of an intersection type) and is likely unrelated.

**Comment by ematipico** [comment]
@liorp, please remove or hide all those AI-generated comments. Update your issue by sharing a playground, and limit yourself to highlighting what's a false positive or false negative. That's all we need.

**Comment by biome-cookie** [comment]
Hello @liorp, please provide a minimal reproduction. You can use one of the following options:

- Provide a link to [our playground](https://biomejs.dev/playground), if it's applicable.
- Provide a link to GitHub repository. To easily create a reproduction, you can use our interactive CLI via `npm create @biomejs/biome-reproduction`

Issues marked with `S-Needs repro` will be **closed** if they have **no activity within 3 days**.

**Comment by liorp** [comment]
Correction to my previous comment: the code path I pointed at is **not compiled**,
so please disregard that part. The behavioural analysis and the reproductions
above still hold, but the file references were wrong.

`crates/biome_js_type_info/src/lib.rs` declares:

```rust
mod builders;
mod format_inferred_type_info;
mod format_type_info;
pub(crate) mod generated;
mod globals;
mod globals_builder;
pub(crate) mod globals_ids;
mod inferred_type;
pub mod interned_types;
mod local_inference;
pub mod resolved;
mod return_type_relation;
mod stringification;
mod type_data;
mod type_store;
mod type_transform;
mod type_traversal;
```

There is no `mod flattening;`, `mod conditionals;`, `mod resolver;`, or
`mod helpers;`, so `src/flattening/`, `src/flattening.rs`, `src/conditionals.rs`,
`src/resolver.rs`, and `src/helpers.rs` are not part of the crate. I confirmed
this by appending `this is not valid rust;` to
`src/flattening/expressions.rs` - `cargo check -p biome_js_type_info` still
succeeds. You may want to delete those files; they read as current and cost me a
wrong diagnosis.

The live path for a call expression is:

1. `biome_module_graph/src/db/type_inference/expressions.rs:114`

   ```rust
   RawTypeofExpression::Call(expression) => {
       let callee = self.resolve(&expression.callee);
       Some(self.resolve_call_expression(callee, &expression.arguments))
   }
   ```

2. `resolve_call_expression` (same file, line 370) calls
   `infer_call_expression_return_type_from_args(self.db, callee, &args)`
3. `biome_module_graph/src/db/queries/type_inference/calls/implementation.rs:66`
   -> `infer_function_call_type` (line 91)
4. -> `infer_function_return_type` (line 1674)
5. -> `infer_generic_return_type(db, function, *ty, args)` (line 1681)

The conclusion is unchanged and the naming makes it explicit: every step threads
`args` and nothing else, so type parameters are bound from the argument types
alone. There is no channel for call-site type arguments, because
`TypeofCallExpression` (`biome_js_type_info/src/type_data.rs:1340`) has only
`callee` and `arguments`.

Two things that may help whoever picks this up:

- `substitutions_for_instance` (`type_inference/lookup.rs:463`) already takes an
  `inherited: &[InferredTypeSubstitution<'db>]` parameter, and
  `implementation.rs:101` passes `&[]`. That looks like the natural seam for
  substitutions coming from explicit type arguments.
- `biome_js_type_info/tests/type_data.rs:67` asserts
  `size_of::<TypeofCallExpression>() == 32` with "The size shouldn't go higher".
  Adding a `Box<[TypeReference]>` field takes it to 48, so if that budget is firm
  the type arguments would need to be carried some other way.

I stopped short of a PR because of that size budget and because the substitution
seam is a design call I would rather not guess at. Happy to implement whichever
direction you prefer, or to test a patch against the repro above.

**Comment by ematipico** [comment]
What the hell man... open a new issue. blocking
